test/logic/models/content_test.dart

```diff
@@ -0,0 +1,297 @@
+        (validSong.copyWith({'id': 18, 'track': null}), songWith(id: 18, track: null)),
+        (validSong.copyWith({'id': 19}).copyWithout(['track']), songWith(id: 19, track: null)),
+        (validSong.copyWith({'id': 20, 'dateAdded': -1}), songWith(id: 20, dateAdded: -1)),
+        (validSong.copyWith({'id': 21, 'dateModified': -1}), songWith(id: 21, dateModified: -1)),
+        (validSong.copyWith({'id': 22, 'dateModified': null}), songWith(id: 22, dateModified: songWith().dateAdded)),
+        (
+          validSong.copyWith({'id': 23}).copyWithout(['dateModified']),
+          songWith(id: 23, dateModified: songWith().dateAdded)
+        ),
+        (validSong.copyWith({'id': 24, 'duration': -1}), songWith(id: 24, duration: -1)),
+        (validSong.copyWith({'id': 25, 'size': -1}), songWith(id: 25, size: -1)),
+        (validSong.copyWith({'id': 26, 'size': null}), songWith(id: 26, size: null)),
+        (validSong.copyWith({'id': 27}).copyWithout(['size']), songWith(id: 27, size: null)),
+        (validSong.copyWith({'id': 28, 'filesystemPath': ''}), songWith(id: 28, filesystemPath: '')),
+        (validSong.copyWith({'id': 29, 'filesystemPath': null}), songWith(id: 29, filesystemPath: null)),
+        (validSong.copyWith({'id': 30}).copyWithout(['filesystemPath']), songWith(id: 30, filesystemPath: null)),
+        (
+          validSong.copyWith({'id': 31, 'isFavoriteInMediaStore': null}),
+          songWith(id: 31, isFavoriteInMediaStore: false)
+        ),
+        (
+          validSong.copyWith({'id': 32}).copyWithout(['isFavoriteInMediaStore']),
+          songWith(id: 32, isFavoriteInMediaStore: false)
+        ),
+        (validSong.copyWith({'id': 33, 'generationAdded': -1}), songWith(id: 33, generationAdded: -1)),
+        (validSong.copyWith({'id': 34, 'generationAdded': null}), songWith(id: 34, generationAdded: null)),
+        (validSong.copyWith({'id': 35}).copyWithout(['generationAdded']), songWith(id: 35, generationAdded: null)),
+        (validSong.copyWith({'id': 36, 'generationModified': -1}), songWith(id: 36, generationModified: -1)),
+        (validSong.copyWith({'id': 37, 'generationModified': null}), songWith(id: 37, generationModified: null)),
+        (
+          validSong.copyWith({'id': 38}).copyWithout(['generationModified']),
+          songWith(id: 38, generationModified: null)
+        ),
+      ];
+      final propertiesThatCanBeMissing = [
+        'album',
+        'albumId',
+        'genre',
+        'genreId',
+        'track',
+        'dateModified',
+        'size',
+        'filesystemPath',
+        'isFavoriteInMediaStore',
+        'generationAdded',
+        'generationModified',
+      ];
+      final invalidSongs = [
+        ...validSong.copyWithout(propertiesThatCanBeMissing).subSets(),
+        ...validSong
+            .withWrongTypes()
+            .whereNot((map) => propertiesThatCanBeMissing.any((property) => map[property] == null))
```

Comment:
What does it do?

Reply:
Generally it's hard to understand what is the output of these function calls, how they work, and why we use them

Reply:
> Generally it's hard to understand what is the output of these function calls, how they work, and why we use them

I agree.

The point of this is to automatically create all possible bad variations of inputs to the `parse` function.
There is two types of problems that this covers:
* Missing data: The `subSets` function gives us variations of the `validSong` map (without the values that are allowed to be null) with one or more missing elements. We generate all possible subsets instead of just checking if it's ok if individual elements are missing because sometimes we fall back to another element in that case (like `Song.dateModified` falling back to `Song.dateAdded`).
* Invalid data types: The `withWrongTypes` gives us variations of the `validSong` map where each element of the map is a different type from the expected one. For example, the `title` property has a type `String`, so `withWrongTypes` will generate one variation where the `title` is `null`, one where the `title` is `true`, one where it's `0` and so on. We only generate variations where one of the elements is changed and don't bother with variations where multiple element types have changed because there would be a large amount of variations and because I don't believe it is very relevant.

I agree that this is not very easy to parse from the code, but I'm not quite sure how to improve the situation other than adding comments and maybe changing the name of `withWrongTypes` and `subSets`. Do you have any recomendations?

Reply:
I've added some helper extension methods to improve the clarity of the tests.

Reply:
I think splitting the tests into the valid and invalid cases improves clarity a lot.